### PR TITLE
Add shadows behind a render-menu flag

### PR DESCRIPTION
**Pull request by zzmp**
Adds shadows behind a menu-item (off by default). Also fixes some shadow-related issues:
- Shadows no longer hold references to old scene items (caused crash on location change)
- Keylight frustum now culls correctly
- Head is rendered, not cauterized, as a shadow
- A light along (0, -1, 0) will retain its orientation (orientation calculation edge-case)
- Extra render buffers are no longer used for debugging

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: ff425f40aec6b1258bcaba856473281ca4bb28d8 available at: https://gist.github.com/5956fa91ef59980cd953<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3c80002dfe08d8a397e1a1d2cb87e50444fc2c25 available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 387a14b6b121f6c46afc068e07d9bfede3a53ff1 available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c794ae8da28a1717991be7d5a4dbc775663d162d available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3445b7fd2c19cafde66e6fc26117916ed3ba85cc available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0c798c06a96ae6036343e48176abc9c0dab40ca8 available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0c798c06a96ae6036343e48176abc9c0dab40ca8 available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by cozza13** [comment]
@zzmp I am putting this back out of QA ready because it looked like something about the build failed

**Comment by zzmp** [comment]
@cozza13 @samcake do you know why the OSX build is failing? I restarted it once already - I'm not sure what is going wrong.

**Comment by cozza13** [comment]
I have no idea

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0c798c06a96ae6036343e48176abc9c0dab40ca8 available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6859.dmg

**Comment by zzmp** [comment]
Strange. It failed twice in a row, and then it worked. All on the same build.

File: libraries/render/src/render/Context.h [hunk 1]
```diff
@@ -97,6 +98,7 @@ class RenderContext {
     bool _drawHitEffect;
     bool _occlusionStatus { false };
     bool _fxaaStatus = { false };
+    bool _shadowMapStatus = { false };
```
Comment (samcake) [inline libraries/render/src/render/Context.h]: here you use = and the racket notation, just {} should be good :)

File: libraries/render-utils/src/skin_model_shadow.slf [hunk 2]
```diff
@@ -15,6 +15,6 @@
 layout(location = 0) out vec4 _fragColor;
 
 void main(void) {
-    // stencil in solid color for debugging
-    _fragColor = vec4(1.0, 0.0, 0.0, 1.0);
+    // pass-through to set z-buffer
+    _fragColor = vec4(1.0, 1.0, 1.0, 0.0);
```
Comment (samcake) [inline libraries/render-utils/src/skin_model_shadow.slf]: need to be 1.0f not 1.0

File: libraries/render-utils/src/model_shadow.slf [hunk 3]
```diff
@@ -15,6 +15,6 @@
 layout(location = 0) out vec4 _fragColor;
 
 void main(void) {
-    // stencil in solid color for debugging
-    _fragColor = vec4(0.0, 0.0, 1.0, 1.0);
+    // pass-through to set z-buffer
+    _fragColor = vec4(1.0, 1.0, 1.0, 0.0);
```
Comment (samcake) [inline libraries/render-utils/src/model_shadow.slf]: need to be 1.0f

File: libraries/render-utils/src/Shadow.slh [hunk 4]
```diff
@@ -60,44 +60,39 @@ float fetchShadow(vec3 shadowTexcoord) {
...
-            fetchShadow(shadowTexcoord.xyz + shadowScale * vec3(samples[7], 0.0))
-        ));
-    }
-
     return shadowAttenuation;
 }
 
 float evalShadowAttenuation(vec4 position) {
 	vec4 shadowTexcoord = evalShadowTexcoord(position);
+    if (shadowTexcoord.x < 0.0 || shadowTexcoord.x > 1.0 ||
```
Comment (samcake) [inline libraries/render-utils/src/Shadow.slh]: use 1.0f not 1.0
i think the coding guidelines say it s ok for 0.0...

File: libraries/render-utils/src/Shadow.slh [hunk 5]
```diff
@@ -60,44 +60,39 @@ float fetchShadow(vec3 shadowTexcoord) {
...
 
+    // Offset for efficient PCF, see http://http.developer.nvidia.com/GPUGems/gpugems_ch11.html
+    vec2 offset = pcfRadius * step(fract(position.xy), vec2(0.5, 0.5));
+
     float shadowAttenuation = (0.25 * (
-        fetchShadow(shadowTexcoord.xyz + shadowScale * vec3(samples[0], 0.0)) +
-        fetchShadow(shadowTexcoord.xyz + shadowScale * vec3(samples[1], 0.0)) +
-        fetchShadow(shadowTexcoord.xyz + shadowScale * vec3(samples[2], 0.0)) +
-        fetchShadow(shadowTexcoord.xyz + shadowScale * vec3(samples[3], 0.0))
+        fetchShadow(shadowTexcoord.xyz + shadowScale * vec3(offset + PCFkernel[0], 0.0)) +
```
Comment (samcake) [inline libraries/render-utils/src/Shadow.slh]: use 0.0f not 0.0

File: libraries/render-utils/src/Shadow.slh [hunk 6]
```diff
@@ -60,44 +60,39 @@ float fetchShadow(vec3 shadowTexcoord) {
...
+    vec2(0.5, -1.5)
 );
 
-float evalShadowAttenuationSampling(vec4 shadowTexcoord) {
+float evalShadowAttenuationPCF(vec4 position, vec4 shadowTexcoord) {
+    float pcfRadius = 3.0;
 	float shadowScale = getShadowScale();
 
+    // Offset for efficient PCF, see http://http.developer.nvidia.com/GPUGems/gpugems_ch11.html
+    vec2 offset = pcfRadius * step(fract(position.xy), vec2(0.5, 0.5));
```
Comment (samcake) [inline libraries/render-utils/src/Shadow.slh]: use 0.5f not 0.5

**Comment by ZappoMan** [comment]
@zzmp - you need to update this PR with a pull from upstream master, or rebase as you see fit so that it can be merged.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dee8e0ba9db0c3c9efecb56642a9c4d2ef009053 available at: https://gist.github.com/2e0ad6db1ada157e0a60<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6859.dmg

**Comment by cozza13** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: b62fc505574230d11e587750aa92af1972890f1c available at: https://gist.github.com/5956fa91ef59980cd953<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6859.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6859.dmg

**Comment by thoys** [comment]
I will thank you!

**Comment by OmegaHeron** [comment]
On Windows 10 with Nvidia GTX660M if I enable shadows it seems they're offset vertically higher than expected - at first I thought it was doing nothing then caught glimpse of darkness.  Placed a cube entity - scaled up to "find" shadow and it's there.

hifi-snap-by-omegaheron-on-2016-01-22_14-40-33

**Comment by OmegaHeron** [comment]
And Mac OS X 10.11.1 Intel HD4000

hifi-snap-by-omegaheron-on-2016-01-22_11-46-02

**Comment by OmegaHeron** [comment]
Can view this example at;

hifi://heron/2099.03,15299.3,2055.98/-0.000152603,0.912939,6.82215e-05,0.408095

**Comment by zzmp** [comment]
@OmegaHeron, thanks for the link. The keyLight in this scene is canted up, such that shadows will be projected up, so this looks like expected behavior.

At this stage in shadows, it is based solely on the keyLight (you can check it in scripts through `Scene.stage.keyLight.direction.x/y/z`). You can try adjusting it and making sure the shadow changes.

**Comment by OmegaHeron** [comment]
@zzmp - Thanks - then within that criteria of expected behavior I'll say this works on Windows/OS X - seems to have no impact upon recently added AO effect and no excess FPS drop.

**Comment by zzmp** [comment]
@OmegaHeron I'm surprised at no excess FPS drop. Do you mean that *without* shadows it runs at the same FPS, or *with* shadows?

**Comment by OmegaHeron** [comment]
With shadows on compared to off I don't see a major change in FPS - a good thing.  I base this upon previous Interface builds with shadows where I remember FPS penalty being greater.  Of course it's not as good on Mac with HD4000, but, that's to be expected for its (lack of) performance.

**Pull request merged**